pkg/kubelet/kubelet.go

```diff
@@ -2952,6 +2969,87 @@ func (kl *Kubelet) fastStatusUpdateOnce() {
+// 'kl.maxCheckpointsPerContainer'.
+func (kl *Kubelet) handleMaxCheckpointsPerContainer(podUID types.UID, podFullName, containerName string, options *runtimeapi.CheckpointContainerRequest) error {
+
+	if kl.maxCheckpointsPerContainer <= 0 {
+		// This cannot really happen as the kubelet config parser will
+		// error out early if this is <= 0. This additional error
+		// handling is mainly here to make sure tests are working.
+		return errors.New("maxCheckpointsPerContainer must be greater than 0")
+	}
+
+	// Lock while accessing the checkpoints on disk to avoid chaos
+	// if multiple threads would do it at the same time.
+	kl.maxCheckpointsPerContainerLock.Lock()
+	defer kl.maxCheckpointsPerContainerLock.Unlock()
+
+	checkpointDirectoryPath := filepath.Join(
+		kl.getCheckpointsDir(),
+		string(podUID),
+	)
+
+	checkpointDirectory, err := os.ReadDir(checkpointDirectoryPath)
+	if err != nil {
+		return fmt.Errorf("error reading checkpoint archive directory %s: %w", checkpointDirectoryPath, err)
+	}
+
+	regexString := fmt.Sprintf(`checkpoint-%s-%s-\d+-\d{%d}.tar`, podFullName, containerName, int(math.Log10(float64(math.MaxInt32))+1))
+
+	re, err := regexp.Compile(regexString)
+	if err != nil {
+		return fmt.Errorf("error compiling regex %s: %w", regexString, err)
+	}
+
+	var checkpointArchives = []fs.FileInfo{}
+	for _, archive := range checkpointDirectory {
+		// only select checkpoint archives which have been created from the same
+		// namespace/pod/container combination
+		if re.MatchString(archive.Name()) {
+			archiveFileinfo, err := archive.Info()
+			if err != nil {
+				return fmt.Errorf("error getting file info %s:%w", archive.Name(), err)
+			}
+			checkpointArchives = append(checkpointArchives, archiveFileinfo)
+		}
+	}
+
+	checkpointArchivesCounter := len(checkpointArchives)
+	if checkpointArchivesCounter <= int(kl.maxCheckpointsPerContainer) {
+		// Nothing to clean up, return early.
+		return nil
+	}
+
+	sort.Slice(checkpointArchives, func(i, j int) bool {
+		return checkpointArchives[i].Name() < checkpointArchives[j].Name()
+	})
+
+	klog.InfoS(
+		"More than maxCheckpointsPerContainer",
+		"container",
+		containerName,
+		"numberOfCheckpoints",
+		checkpointArchivesCounter,
+		"maxCheckpointsPerContainer",
+		kl.maxCheckpointsPerContainer,
+	)
+	for _, archive := range checkpointArchives[:checkpointArchivesCounter-int(kl.maxCheckpointsPerContainer)] {
+		archiveWithPath := filepath.Join(checkpointDirectoryPath, archive.Name())
+		klog.InfoS("Deleting container checkpoint archive", "container", containerName, "archive", archiveWithPath)
+		err = os.Remove(archiveWithPath)
+		if err != nil {
+			// If the removal fails the checkpoint archive might have been
+			// deleted externally. This is not fatal but unexpected.
+			// Logging it as an error, but not returning an error.
+			klog.ErrorS(err, "Removal of checkpoint archive failed", "container", containerName, "archive", archiveWithPath)
```

Comment:
If this error is to be a warning, then perhaps it makes sense to log it at that level?

Reply:
There is no `WarnS()` function but according to some discussion `InfoS()` is the correct way to do it. I changed it to `InfoS()`.

Reply:
Ah, now I remember. `ErrorS()` is used to get the information from the failed `os.Remove()` which gives us an `err`.